hicommonwealth/commonwealth
# Create Outbox

## Link to Issue
Closes: #6924 

## Description of Changes
- Creates the partitioned `Outbox` table.
- Created the `outbox-archive.ts` script which backs up all partitioned child tables containing data from the previous month (or more).
- Executes the `pg_partman` `run_maintenance()` function as part of the database cleaner flow. This executes at least once a day and will create new child tables 2 months in advance and drop child tables older than 6 months.

## Test Plan
If you use Docker Postgres:
- Run `docker compose up` as usual
- Run `yarn db-all`

If you use native Postgres:
- Install [`pg_partman`](https://github.com/pgpartman/pg_partman?tab=readme-ov-file#installation)
- Run `yarn db-all`

Testing the archival script is a little more complicated:
- Deploy this branch/PR to Frick
- Copy production AWS credentials to Frick
- Execute the following script to populate the Outbox table with some test data:
```
DO $$
DECLARE
  start_date TIMESTAMP WITH TIME ZONE;
  end_date TIMESTAMP WITH TIME ZONE;
BEGIN
  -- Set the start and end date for the 3 months period
  start_date := CURRENT_DATE - INTERVAL '3 months';
  end_date := CURRENT_DATE;

  -- Generate 1,000 rows of test data, with dates spread across the last 3 months
  INSERT INTO "Outbox" (event_name, event_payload, relayed, created_at, updated_at)
  SELECT
    'Event_' || LPAD((row_number() OVER (ORDER BY s)::text), 4, '0'), -- Generate a prefixed event name with padding for uniqueness
    jsonb_build_object('key', MD5(random()::text)), -- Generate a random JSONB object
    TRUE,
    s, -- Use the generated timestamp for created_at
    s  -- Use the same generated timestamp for updated_at
  FROM generate_series(start_date, end_date, '1 day'::interval / (1000 / 90)) AS s
  LIMIT 1000;
END $$;
```
- Start a Heroku one-off dyno: `heroku run bash -a commonwealth-frick`
- Execute `yarn archive-outbox`
  - If the tables are already backed up the script will skip them. If you empty the bucket and rerun the script, the new dumps should be uploaded.


## Deployment Plan

1. Execute `yarn archive-outbox` with Heroku scheduler once a week

## Other Considerations

- Devs who don't use the Docker compose file will need to install the [`pg_partman` extension](https://github.com/pgpartman/pg_partman?tab=readme-ov-file#installation) separately.
